File: 2.0/8/A.cpp
#include <iostream>
#include <string>
#include <sstream>

using namespace std;


struct Node {
    int key;
    Node* left;
    Node* right;
};

class Tree {
    private:
    Node* root;
    bool flag = false;

    Node* add(Node* node, int n) {
        if (!node) {
            return new Node{n, nullptr, nullptr};
        }
        if (n == node->key){
            flag = true;
            return node;
        } else if (n < node->key) {
            node->left = add(node->left, n);
        } else if (n > node->key){
            node->right = add(node->right, n);
        }
        return node;
    }


    void search(Node* node, int n) {
        if (!node){
            return;
        }
        if (n == node->key){
            flag = true;
            return;
        } else if (n < node->key) {
            search(node->left, n);
        } else if (n > node->key) {
            search(node->right, n);
        }
        return;
    }

    void print(Node* node, string& res, int lvl) {
        if (!node) return;

        print(node->left, res, lvl + 1);
        if (node->left) res += "\n";

        res += string(lvl, '.');
        res += to_string(node->key);

        if (node->right) res += "\n";
        print(node->right, res, lvl + 1);
    }

    public:
    Tree() : root(nullptr){}


    string add(int n){
        flag = false;
        root = add(root, n);
        return flag ? "ALREADY" : "DONE";
    }

    string search(int n){
        flag = false;
        search(root, n);
        return flag ? "YES" : "NO";
    }

    string tostring(){
        string res;
        print(root, res, 0);
        return res;
    }
};


int main() {
    ios::sync_with_stdio(false), cin.tie(0), cout.tie(0);
    Tree tree;

    string line;
    getline(cin, line);
    while(!line.empty()) {
        stringstream ss(line);

        string cmd;
        int n;
        ss >> cmd >> n;
        if (cmd == "PRINTTREE") {
            cout << tree.tostring() << endl;
        } else if (cmd == "ADD") {
            cout << tree.add(n) << endl;
        } else {
            cout << tree.search(n) << endl;
        }
        getline(cin, line);
    }

    return 0;
}

File: 2.0/8/A.py
from sys import stdin

class Node:
    def __init__(self, key: int, flag: bool = False):
        self.left = None
        self.right = None
        self.key = key
        self.flag = flag

class Tree:
    def __init__(self):
        self.root = None
    
    def _add(self, node: Node, n: int) -> Node:
        if not node:
            return Node(n)
        if n == node.key:
            self.flag = True
            return node
        elif n < node.key:
            node.left = self._add(node.left, n)
        else:
            node.right = self._add(node.right, n)
        return node
    
    def _search(self, node: Node, n: int) -> None:
        if not node:
            return
        if n == node.key:
            self.flag = True
            return
        elif n < node.key:
            self._search(node.left, n)
        else:
            self._search(node.right, n)
        return
    
    def tostring(self, node: Node, res: str, lvl: int) -> str:
        if node.left:
            res = self.tostring(node.left, res, lvl + 1) 
            res += "\n"

        res += lvl * "."
        res += str(node.key)

        if node.right:
            res += "\n"
            res = self.tostring(node.right, res, lvl + 1)

        return res
    
    
    def add(self, n: int) -> str:
        self.flag = False
        self.root = self._add(self.root, n)
        return "ALREADY" if self.flag else "DONE"
    
    def search(self, n: int) -> str:
        self.flag = False
        self._search(self.root, n)
        return "YES" if self.flag else "NO"
    
    def __str__(self):
        ans = ""
        return self.tostring(self.root, ans, 0)




tree = Tree()
for line in stdin:
    line = line.strip().rstrip().split()

    if line[0] == "PRINTTREE":
        print(tree)
    elif line[0] == "ADD":
        print(tree.add(int(line[1])))
    elif line[0] == "SEARCH":
        print(tree.search(int(line[1])))
        